### ISSUE TITLE
shpool session creates with pwd as home directory not on curent path

### ISSUE THREAD

**Issue by rinkugoel** [shell-pool/shpool#47]
**What happened**
When i create the new session it takes me to home path ~/ 

**What I expected to happen**
I should remain at the same directory path I was preset.

**To Reproduce**
1. go to some path in your system other then home directory 
2. create session using shpool
3. you will be on home directory in the shpool. That was not the case with tmux and i liked that much

**Version info**
Run `shpool version` and paste the output here

**Logs**
Run `journalctl --user -u shpool` to pull logs related to shpool. You will likely want to use a pastebin to post the logs, and you may need to trim the output with `tail`.

**Comment by ethanpailes** [comment]
This should be pretty simple to address. We can add a config option to allow people to switch between the different behavior.

I'm not sure if we should break compatibility to change this though. To get a feel for what people want, give this comment a ❌ react if you want to see compatibility broken to make this the new default and a 🛑 react if you want to see the current behavior stay the same with just an option to choose different behavior if you want.

**Comment by ethanpailes** [comment]
If we retain the current behavior by default, I think a good flag name to switch the behavior would be `-h`/`--here`. We can also add a config option to make it the default.

**Comment by justinsb** [comment]
FYI, I don't think I'm able to give arbitrary reactions.  My vote would be ❌ (make the breaking change), but we may want to change the options to 👍  and 👎 

Is there a workaround (like a way to send arbitrary commands?).  The best I've come up with is: 

```
shpool attach testbash --cmd 'bash -c "cd ~/bin; bash"'
```

That seems to work but feels a "little" hacky!

**Comment by Aetf** [comment]
I haven't checked, but does shpool inherit full client side env vars on creating new session? PWD is just one of such env vars (well, a bit more special then just the PWD env var, but anyway). This is related to https://github.com/shell-pool/shpool/issues/111#issuecomment-2302955472, and I think we should have a consistent story around env vars/pwd.

**Comment by ethanpailes** [comment]
shpool doesn't inherit the full env, but there is a configurable allowlist with a short list of defaults for variables that will get forwarded from the client environment and injected into the daemon. The ability to just forward everything has come up before and wouldn't be too hard to add, but I think we wanted to wait until someone articulated a good usecase before adding it.

**Comment by jul-sh** [comment]
> shpool doesn't inherit the full env, but there is a configurable allowlist with a short list of defaults for variables that will get forwarded from the client environment and injected into the daemon. The ability to just forward everything has come up before and wouldn't be too hard to add, but I think we wanted to wait until someone articulated a good usecase before adding it.

how can one configure this allow list? In general I'd think inherting all ENVs would be the most intuitive behavior for new sessions.

**Comment by jul-sh** [comment]
> forward_env

Thanks! I can't get it to work for PWD; anything obvious I'm doing wrong?


~/.config/shpool/config.toml
```
session_restore_mode = { lines = 1000 }
forward_env = ["PWD"]
```

**Comment by jul-sh** [comment]
> Since your shell will immediately set PWD, how are you using it before that happens?

unsure what you mean; basically the goal is that I run "shpool attach foo" in a given dir, and then the shpool shell is also in this dir.

**Comment by ethanpailes** [comment]
Oh, that won't be done via foward_env. We'll need to write some new code for that.

**Comment by jul-sh** [comment]
ah okay! I tried hacking around to get this by using the --cmd flag and an alias to accomplish this; but it gets a bit complicated since the --cmd flag replaces the shell invocation entirely. How hard do you think it'd be to implement this?

**Comment by ethanpailes** [comment]
I don't think this would be too hard. If you are interested in taking a swing at implementing it, I could link you to some points in the code that would be good starting points.

If you just want to hack around the problem, what you could put `foward_env = ["SHPOOL_CD_TO"]` in your config, then write a wrapper function to invoke shpool with that does `SHPOOL_CD_TO=$(pwd) shpool attach -f <session-name>` and add a stansa to your ~/.bashrc that checks if SHPOOL_CD_TO is set, and if so, changes to that directory.

**Comment by jul-sh** [comment]
If you could provide some code pointers I'd be interested in giving it a shot. 

I couldn't make the workaround you mentioned work, since in my testing the env wasn't passed through (despite `foward_env = ["SHPOOL_CD_TO"]` being set) and since the zshrc doesn't run again upon entering the shpool shell. 

In general it'd be cool to have a hook that permits users to specify a command that should be run when entering the shpool shell.

**Comment by ethanpailes** [comment]
> I couldn't make the workaround you mentioned work, since in my testing the env wasn't passed through (despite foward_env = ["SHPOOL_CD_TO"] being set) and since the zshrc doesn't run again upon entering the shpool shell.

This would only work on shell startup, not reattach. There is no way for us to update environment variables on the fly (well technically there kinda is which is what I wrote https://github.com/shell-pool/subprocess-inject-env, but it only works for bash since the other shells I tested seem to cache environment variables, but even so I haven't actually integrated that crate yet). Changing directory on reattach would be even harder because we would need to figure out some way to issue a `cd` command to the shell. At first, this might sound simple enough, but the trouble is that we have no idea what state the shell is in when we are doing the reattach. It might just be happily sitting there ready for a command to be issued, but it could also be in the middle of a vim or emacs sessions, which would get mangled if we tried to inject a command on reattach.

In terms of places to look to implement this, we spawn new shells [here](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/daemon/server.rs#L662), and set the spawn directory for the new shell [here](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/daemon/server.rs#L705), so your goal will be modifying that line with the attach invocation directory rather than the user's home directory. To do that, you'll want to add a new optional field to the [AttachHeader](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/shpool-protocol/src/lib.rs#L192), and a new optional command line flag to the [attach subcommand](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/lib.rs#L127). `--in-pwd` seems like it would make sense to me, but we can bikeshed that. You'll have to plumb the flag value down into the [attach::run](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/lib.rs#L360) routine, and then put it in the AttachHeader you've modified. Then in the daemon you'll need to modify [handle_attach](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/daemon/server.rs#L206) to factor the new field into its shell creation routine, which is called [here](https://github.com/shell-pool/shpool/blob/4374c9a85e87b2a13f9a59017e9d4b8dc6bbeead/libshpool/src/daemon/server.rs#L309).

**Comment by ethanpailes** [comment]
You could also consider adding a config option that makes this the default behavior.

**Comment by tobwen** [comment]
My two cents: Doesn't `-h` collide with `-h, --help`?

**Comment by ethanpailes** [comment]
Yeah you're right. We could just drop the short flag or maybe there is some other name that would also work.

**Comment by ethanpailes** [comment]
Yeah, I like `-d/--dir`
